text_machina/src/generators/mixcase.py

```diff
@@ -78,6 +89,195 @@ def _pack(self, generations: List[str], **kwargs) -> Dataset:
+            for idx, position in enumerate(positions):
+                start_pos = len("".join(sentences_with_rewrites[:position]))
+                end_pos = len("".join(sentences_with_rewrites[: position + 1]))
+                generated_labels.append(
+                    LabeledSpan(
+                        start=start_pos,
+                        end=end_pos,
+                        label=DetectionLabels.GENERATED.value,
+                    ).model_dump()
+                )
+
+            # Mark human spans
+            human_labels = []
+            prev_generated_position = 0
+            for idx in range(len(generated_labels)):
+                generated_label = generated_labels[idx]
+                print(prev_generated_position, generated_label["start"])
+                if prev_generated_position < generated_label["start"]:
+                    human_labels.append(
+                        LabeledSpan(
+                            start=prev_generated_position,
+                            end=generated_label["start"],
+                            label=DetectionLabels.HUMAN.value,
+                        ).model_dump()
+                    )
+                prev_generated_position = generated_label["end"]
+
+            # Add human label if the end of the text has not been
+            # reached by the last generated span
+            last_generated_position = generated_labels[-1]["end"]
+            if last_generated_position < len(sample_text):
+                human_labels.append(
+                    LabeledSpan(
+                        start=last_generated_position,
+                        end=len(sample_text),
+                        label=DetectionLabels.HUMAN.value,
+                    ).model_dump()
+                )
+            # Merge generated and human spans
+            # Join generated labels with human labels and merge overlappings.
+            merged_labels = sorted(
+                human_labels + generated_labels, key=lambda span: span["start"]
+            )
+            sample_labels = []
+            while len(merged_labels):
+                current_span = merged_labels.pop(0)
+                current_label = current_span["label"]
+                spans_with_same_label = []
+                while (
+                    len(merged_labels)
+                    and merged_labels[0]["label"] == current_label
+                ):
+                    spans_with_same_label.append(merged_labels.pop(0))
+
+                if not spans_with_same_label:
+                    sample_labels.append(current_span)
+                else:
+                    sample_labels.append(
+                        LabeledSpan(
+                            start=current_span["start"],
+                            end=spans_with_same_label[-1]["end"],
+                            label=current_label,
+                        ).model_dump()
+                    )
+
+            prev_sample += len(positions)
```

Comment:
A bit of context:

- `self.extractor.workspace["positions"]` is a list of lists, where each inner list **i** has the positions that has to be replaced in the sample **i**. So, `positions` is one of that inner lists, that has as many elements as sentences to be replaced in a sample.
- `generations` is just a list, there is no notion in it about what completion should be used to replace sentences in each sample.

We can't work with any other data structure for `generations` because doing so now will change so much the logic for all the tasks. That's the reason why I'm using the `prev_sample`. It starts by 0 (first sample), then do:

```python
rewrittens = generations[prev_sample : prev_sample + len(positions)]
```

to pick the generated sentences corresponding to the sample 0, and at the end of the loop, it is increased by `len(positions)`, meaning that we should move to the generations for the next sample.

If an extractor does correctly its work, the number of items in `self.extractor.workspace["positions"]` (flattened) must match the number of generations, and then there will be no indexing errors.